src/systems/filecoin_blockchain/chainsync/_index.md

```diff
@@ -1,58 +1,490 @@
+
+- Blocks can be validated in progressive stages, in order to minimize resource expenditure.
+- Validation computation is considerable, and a serious DOS attack vector.
+- Secure implementations must carefully schedule validation and minimize the work done by pruning blocks without validating them fully.
+- **Progressive Stages of Block Validation**
+  - _(TODO: move this to blockchain/Block section)_
+  - **BV0 - Syntactic Validation**: Validate data structure packing and ensure correct typing.
+  - **BV1 - Light Consensus State Checks**: Validate `b.ChainWeight`, `b.ChainEpoch`, `b.MinerAddress`, are plausible (some ranges of bad values can be detected easily, especially if we have the state of the chain at `b.ChainEpoch - consensus.LookbackParameter`. Eg Weight and Epoch have well defined valid ranges, and `b.MinerAddress`
+  must exist in the lookback state). This requires some chain state, enough to establish plausibility levels of each of these values. A node should be able to estimate valid ranges for `b.ChainEpoch` based on the `LastTrustedCheckpoint`. `b.ChainWeight` is easy if some of the relatively recent chain is available, otherwise hard.
+  - **BV2 - Signature Validation**: Verify `b.BlockSig` is correct.
+  - **BV3 - Verify ElectionProof**: Verify `b.ElectionProof` is correct. This requires having state for relevant lookback parameters.
+  - **BV4 - Verify Ancestry links to chain**: Verify ancestry links back to trusted blocks. If the ancestry forks off before finality, or does not connect at all, it is a bad block.
+  - **BV4 - Verify MessageSigs**: Verify the signatures on messages
+  - **BV5 - Verify StateTree**: Verify the application of `b.Parents.Messages()` correctly produces `b.StateTree` and `b.MessageReceipts`
+- These stages can be used partially across many blocks in a candidate chain, in order to prune out clearly bad blocks long before actually doing the expensive validation work.
+
+Notes:
+- in `CHAIN_CATCHUP`, if a node is receiving/fetching hundreds/thousands of `BlockHeaders`, validating signatures can be very expensive, and can be deferred in favor of other validation. (ie lots of BlockHeaders coming in through network pipe, dont want to bound on sig verification, other checks can help dump blocks on the floor faster (BV0, BV1)
+- in `CHAIN_FOLLOW`, we're not receiving thousands, we're receiving maybe a dozen or 2 dozen packets in a few seconds. We receive cid w/ Sig and addr first (ideally fits in 1 packet), and can afford to (a) check if we already have the cid (if so done, cheap), or (b) if not, check if sig is correct before fetching header (expensive computation, but checking 1 sig is way faster than checking a ton). In practice likely that which one to do is dependent on miner tradeoffs. we'll recommend something but let miners decide, because one strat or the other may be much more effective depending on their hardware, on their bandwidth limitations, or their propensity to getting DOSed
+
```

Comment:
We need a note on GC in chainSync. Specifically with the fact that top-X heaviest heads' blocks and graphs should be kept around until past the finality threshold. Others can be discarded, all can be discarded at finality (SHOULD)

Reply:
Also, having a set of diagrams here would be super helpful to illustrate how this all works. We should have a toy example at the top or bottom of this index file to disambiguate intent and show off a lot of things (with code remaining canon)

Reply:
> We need a note on GC in chainSync. Specifically with the fact that top-X heaviest heads' blocks and graphs should be kept around until past the finality threshold. Others can be discarded, all can be discarded at finality (SHOULD)

yeah great point

> Also, having a set of diagrams here would be super helpful to illustrate how this all works. We should have a toy example at the top or bottom of this index file to disambiguate intent and show off a lot of things (with code remaining canon)

what kind of diagrams would you like to see? (rough sketch for me, i'll make them later)

Reply:
addressed (cache)

Reply:
The diagrams that we made when first discussing this in LA. Showing a toy example with say 10 blocks all coming in, then being worked through the pipeline into fringe and then into longest target head. 

So some set of colored blocks being validated and classified over multiple steps allowing us to see the sequence of how "sync" should work.